### PR TITLE
Feature/decorator

### PR DESCRIPTION
**Pull request by MVacante**

### CODE REVIEW

File: src/main/java/tool/MainProgram.java [hunk 1]
```diff
@@ -32,7 +32,7 @@ private MainProgram() {
     public static void main(String[] args) {
         //Just use this project for now (src), will have to change
         //to the target project with the gradle stuff
-        startAnalyse(new String[] {"src"});
+
```
Comment (ericlp) [inline src/main/java/tool/MainProgram.java]: Det här håller jag inte med om. Varför är den borttagen?

File: src/test/java/mocks/decorator/failingpattern/FailingCream.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
+package mocks.decorator.failingpattern;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+/**
+ * Should fail predicate because the class does not set it's component at any point
```
Comment (ericlp) [inline src/test/java/mocks/decorator/failingpattern/FailingCream.java]: Is this comment correct? It does set the component in the first constructor

File: src/test/java/designpatternmocks/singletonpattern/EagerSingletonMock.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    /**
+     * private static method getInstance which returns the instance, should pass tests
+     *
+     * @return The Singleton instance of SingletonMock
+     */
+    public static EagerSingletonMock getInstance() {
+        return instance;
+    }
+
+}
```
Comment (ericlp) [inline src/test/java/designpatternmocks/singletonpattern/EagerSingletonMock.java]: I believe this file is meant to be in a different PR

File: src/main/java/tool/designpatterns/verifiers/multiclassverifiers/decorator/DecoratorPatternInstance.java [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+                                    }
+                                });
+                            });
+                        }
+                    });
+                });
+                // If any elements are left then they are invalid instances
+                concreteComponents.removeAll(identifiedElements);
+                abstractDecorators.removeAll(identifiedElements);
+                concreteDecorators.removeAll(identifiedElements);
```
Comment (ericlp) [inline src/main/java/tool/designpatterns/verifiers/multiclassverifiers/decorator/DecoratorPatternInstance.java]: Are they invalid in that way of an illegal argument or wrongful use a designpattern? If it is the latter, then they should each create a feedback signifying that they are incorrect.

File: src/main/java/tool/designpatterns/verifiers/multiclassverifiers/decorator/DecoratorPatternInstance.java [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+            } else {
+                throw new UnsupportedOperationException("Was not an interface");
```
Comment (ericlp) [inline src/main/java/tool/designpatterns/verifiers/multiclassverifiers/decorator/DecoratorPatternInstance.java]: The only case that this `else` will be called is if the list contains Classes as well as Interfaces, correct? If so, then the class has been marked incorrectly and it should be reported with a `Feedback`.

Comment (ericlp) [reply src/main/java/tool/designpatterns/verifiers/multiclassverifiers/decorator/DecoratorPatternInstance.java]: Or wait, is that handled below?

File: src/main/java/tool/designpatterns/verifiers/multiclassverifiers/decorator/DecoratorVerifier.java [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+
+        if (resultBool.get()) {
+            result = Feedback.getSuccessfulFeedback();
+        } else {
+            String msg = "Interface does not contain any methods!";
+            if (erroringMethod == null) {
+                result = Feedback.getPatternInstanceNoChildFeedback(msg);
+            } else {
+                result = Feedback.getNoChildFeedback(msg, new FeedbackTrace(erroringMethod));
+            }
```
Comment (ericlp) [inline src/main/java/tool/designpatterns/verifiers/multiclassverifiers/decorator/DecoratorVerifier.java]: This method is never called with a MethodDeclaration as an argument and erroringMethod is therefore always null if the result is false. This method could either be simplified or a step may have been missed in some verification.

File: src/main/java/tool/designpatterns/verifiers/multiclassverifiers/decorator/DecoratorVerifier.java [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+     * @return A {@link Feedback} object containing the result
+     */
+    @SuppressWarnings("PMD.LinguisticNaming")
+    private Feedback hasFieldOfType(
+        ClassOrInterfaceDeclaration toTest, ClassOrInterfaceDeclaration type) {
+        Feedback result;
+        AtomicBoolean hasAComponent = new AtomicBoolean(false);
+        toTest.findAll(VariableDeclarator.class).forEach(fieldDeclaration -> {
+            if (fieldDeclaration.getTypeAsString().contains(
+                type.getNameAsString())) { //Check that this works
```
Comment (ericlp) [inline src/main/java/tool/designpatterns/verifiers/multiclassverifiers/decorator/DecoratorVerifier.java]: Why not simply use .equals()? I tested this now and .equals should do the job.
waaaa

File: src/main/java/tool/designpatterns/verifiers/multiclassverifiers/decorator/DecoratorVerifier.java [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+     * @return A {@link Feedback} object containing the result
+     */
+    private Feedback fieldInitializedInConstr(
+        ClassOrInterfaceDeclaration toTest, ClassOrInterfaceDeclaration interfaceName) {
+        Feedback result;
+        AtomicBoolean isInitialized = new AtomicBoolean(true);
+        List<FieldDeclaration> fieldsInClass = new ArrayList<>();
+        toTest.findAll(FieldDeclaration.class).forEach(fieldDeclaration -> {
+            fieldsInClass.add(fieldDeclaration);
+        });
```
Comment (ericlp) [inline src/main/java/tool/designpatterns/verifiers/multiclassverifiers/decorator/DecoratorVerifier.java]:
```suggestion
        List<FieldDeclaration> fieldsInClass = new ArrayList<>(
            toTest.findAll(FieldDeclaration.class));
```

Comment (ericlp) [reply src/main/java/tool/designpatterns/verifiers/multiclassverifiers/decorator/DecoratorVerifier.java]: Intellij suggested this change for me and it hade some other suggestions as well. Use Inteillij well and it can simplify and shorten the code for you.

File: src/test/java/designpatternmocks/singletonpattern/EagerSingletonMock.java (hunk 3 above)
Comment (Darkarca) [reply src/test/java/designpatternmocks/singletonpattern/EagerSingletonMock.java]: File removed in next commit

File: src/test/java/mocks/decorator/failingpattern/FailingCream.java (hunk 2 above)
Comment (Darkarca) [reply src/test/java/mocks/decorator/failingpattern/FailingCream.java]: Comment changed, nice catch! =D

File: src/main/java/tool/MainProgram.java (hunk 1 above)
Comment (Darkarca) [reply src/main/java/tool/MainProgram.java]: We used main to test the program during development before we had real tests written, accidentaly removed startAnalyse, reverted this change in next commit

File: src/main/java/tool/designpatterns/verifiers/multiclassverifiers/decorator/DecoratorVerifier.java (hunk 6 above)
Comment (Darkarca) [reply src/main/java/tool/designpatterns/verifiers/multiclassverifiers/decorator/DecoratorVerifier.java]: During the change to feedback some misinterpretation of the code happened which resulted in this. Resolved the unaccessible block aswell as restructured the code that caused this confusion in the first place

File: src/main/java/tool/designpatterns/verifiers/multiclassverifiers/decorator/DecoratorVerifier.java (hunk 7 above)
Comment (Darkarca) [reply src/main/java/tool/designpatterns/verifiers/multiclassverifiers/decorator/DecoratorVerifier.java]: Equals should do the trick, good suggestion!

File: src/main/java/tool/designpatterns/verifiers/multiclassverifiers/decorator/DecoratorVerifier.java (hunk 8 above)
Comment (Darkarca) [reply src/main/java/tool/designpatterns/verifiers/multiclassverifiers/decorator/DecoratorVerifier.java]: Reasonable change, missed that findAll returns a List rather than some other type, this can probaply be applied elsewhere in the code aswell!

**Review by ericlp — DISMISSED** [review]
Good job! :+1:

**Review by lir108 — DISMISSED** [review]
Looks good

File: src/main/java/tool/designpatterns/verifiers/multiclassverifiers/decorator/DecoratorVerifier.java [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+        }
+        if (isInitialized.get()) {
+            result = Feedback.getSuccessfulFeedback();
+        } else {
+            result = Feedback.getNoChildFeedback(
+                "All constructors did not initialize the Component field",
+                new FeedbackTrace(toTest));
+        }
+        return result;
+    }
```
Comment (ViddeM) [inline src/main/java/tool/designpatterns/verifiers/multiclassverifiers/decorator/DecoratorVerifier.java]: I would really like to see these extracted to a seperate class for reuse in other verifications.

**Review by ViddeM — DISMISSED** [review]
Looks good, please look at my comment tho.

File: src/main/java/tool/designpatterns/verifiers/multiclassverifiers/decorator/DecoratorVerifier.java (hunk 9 above)
Comment (MVacante) [reply src/main/java/tool/designpatterns/verifiers/multiclassverifiers/decorator/DecoratorVerifier.java]: Sounds reasonable! Should we extract them to PatternUtils?

Comment (ericlp) [reply src/main/java/tool/designpatterns/verifiers/multiclassverifiers/decorator/DecoratorVerifier.java]: Would rather see a ClassOrInterfaceUtils as PatternUtils are to be used by Pattern, not classes

**Review by ericlp — APPROVED** [review]

**Comment by ViddeM** [comment]
Should be merged only after symbol solver.

**Review by hulthe — APPROVED** [review]
Look good :+1:

**Review by ViddeM — APPROVED** [review]

**Review by ejohnsson6 — APPROVED** [review]
:+1: :rocket:
